### DISCUSSION TITLE
OpenSSL crashes inside Curl

### DISCUSSION THREAD

**Discussion by Shackster** [openssl/openssl#25212] [Q&A]
Hello Team,
I am using curl 8.7.1
language - C++
I am getting below crash when I send high volume of requests to one of our SSL servers. A dbx stack trace is provided below:

```
(dbx) where
.(??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??) at 0x100000000
look_str_cb@AF19_12() at 0x10063efac
int_cb_LHASH_DOALL_ARG() at 0x100544f58
doall_util_fn@AF15_10() at 0x1005243e8
engine_table_doall() at 0x100544fa8
ENGINE_pkey_asn1_find_str() at 0x10063e8dc
EVP_PKEY_asn1_find_str() at 0x1005a7b88
get_optional_pkey_id() at 0x1007c8c28
IPRA.$ssl_cipher_get_disabled() at 0x1007c9858
ssl_create_cipher_list() at 0x1007caf6c
SSL_CTX_set_cipher_list() at 0x1007c2644
ossl_connect_step1() at 0x1007b9508
ossl_connect_common() at 0x1007bc4f8
Curl_ssl_connect_nonblocking() at 0x1007b0b7c
Curl_http_connect() at 0x10085eb04
IPRA.$multi_runsingle() at 0x100825eec
curl_multi_perform() at 0x100827278
easy_perform() at 0x1007aea10
internal error: unexpected value 120 at line 5214 in file stabstring.c
internal error: unexpected value 120 at line 5214 in file stabstring.c
internal error: unexpected value 120 at line 5214 in file stabstring.c
```

It appears to be crashing deep inside OpenSSL. Non-ssl works fine. The version and OS information is presented below.

Operating System: AIX 7.2

Library and version information:
curl 8.7.1 (powerpc-ibm-aix7.1.5.0) libcurl/8.7.1 **OpenSSL/1.1.1v** zlib/1.2.13 libssh2/1.10.0 nghttp2/1.58.0 OpenLDAP/2.5.16
Release-Date: 2024-03-27
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz NTLM SPNEGO SSL threadsafe UnixSockets

I have a similar issue created for the Curl devs:
https://github.com/curl/curl/issues/14517

Any Ideas on how this could happen? Any and all information and advice are welcome. Thanks in advance!

**Comment by nhorman** [comment]
Just FYI, 1.1.1 is out of support, you should try to recreate this on openssl 3 if you can

That said, I'm not sure what the dbx meaning of this is:
.(??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??, ??) at 0x100000000

I'm sort of assuming its dbx's way of saying I jumped to a function that I have no stabs information for (perhaps an invalid function pointer?)

can you elaborate on what dbx means by that?

**Comment by Shackster** [comment]
It seems to mean what you think it does. It goes on to say "Illegal Instruction" in . at 0x100000000.

**Comment by bernd-edlinger** [comment]
Are you using any openssl engines here?

**Reply by Shackster** [reply]
Do you mean like a specific cipher suite?

**Reply by bernd-edlinger** [reply]
no, I ask in general, if there is an external crypto-engine installed.

**Reply by Shackster** [reply]
I think we are using the one that is built into openssl. Is there a way to accurately verify?

**Reply by nhorman** [reply]
If you provide your openssl config, we should be able to tell, unless you've built something into curl

**Reply by Shackster** [reply]
The curl version that I have was installed from ibm's toolbox webpage without a particular engine in mind. I'm fairly certain I am using openssl's engine. I do however have openssl 3.0 installed separately and have found several openssl config files in different places. Is it possible openssl/libcurl is getting confused? How do I know which config file belongs to what? I imagined prebuilt libraries wouldn't even come with a config file.

**Reply by nhorman** [reply]
run strace on the binary and look for calls to open() that include the name openssl.cnf

**Reply by Shackster** [reply]
strace doesn't seem to accept binaries as parameters on AIX. Is there another way? If I changed all of the profile file names to perhaps openssl.cnf.temp, would usage of openssl throw some kind of error?

**Reply by nhorman** [reply]
you can use truss instead on AIX

**Reply by Shackster** [reply]
This is what came up from the usage of the truss command on the executable with its output piped into a text file and then grepped for the string "openssl.cnf".

kopen("/home/jenkins/j2-edge/workspace/openssl/1.0.2zd@2/openssl-1.0.2zd-pkg_aix64/openssl.cnf", O_RDONLY) Err#2  ENOENT

This seems to fit the bill, but this path does not at all exist on my machine.

**Reply by Shackster** [reply]
Maybe thats the problem?

**Reply by Shackster** [reply]
Any idea on what could cause this? or fix this?

**Reply by nhorman** [reply]
Ostensibly, when you build openssl, you are doing so from a jenkins bot, and the build command configures openssl with the --openssldir option pointing to something in its own workspace.  something like:
```
./Configure --openssldir $WORKSPACE
```
Which would create a default OPENSSLDIR that has a path reasonable for the jenkins system, but no other installation site.  You can override that by setting OPENSSL_CONF on the system you are running on to point to the config file you want to use.

Longer term, you should probably adjust your jenkins build to point to something more universally reasonable

**Reply by Shackster** [reply]
I don't build them. Curl and openssl are prebuild by IBM and I just have them downloaded onto my machine. Could this be the cause for the crash? Should I consider building from source myself?

**Reply by Shackster** [reply]
Perhaps another but uglier option would be to simply create that path and file myself locally. But wouldn't I have to rebuild anyways?

**Comment by nhorman** [comment]
I think bernd's question is a valid one.  From the look of the stack trace, and your description of the dbx display of the ?? symbols, it sounds like look_str_cb attempted to call e->pkey_asn1_meths, and jumped to some random location in ram.  That sounds to me like an external crypto engine that perhaps tore down its engine struct before it was supposed to.  Can you pop up a  frame in your call stack and interrogate the e pointer there to see what we're working with?

**Reply by KristianIvarsson** [reply]
I don't know if it's worth meantioning it, but we've got various crashes when using libcurl with SFTP in a multithreaded environment (without using any common CURL handles or such, i.e. the threads are completely isolated (or at least should be)).

It's on a Windows system and according to the core-dumps it is usually some uninitialized memory access (or such) in libssl or libcrypto code, meaning that it is probably something corrupting the memory but not necessarily in libbssl or libcrypto. If using just one thread (or plain FTP), the behaviour is fine.

We've tried libssh2 1.10 & 1.11, libcurl 8.7 & 8.11, libopenssl 3.2 & 3.4 and almost all combinations

This worked with libssh2 1.10, libcurl 7.8, libopenssl 3.0
